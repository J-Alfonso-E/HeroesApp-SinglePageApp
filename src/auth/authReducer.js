import { types } from '../types/types';

/*const state = {
    name: 'Poncho', 
    logged: true,
};*/

export const authReducer = (state = {}, action) => {
    //console.log(state);
    //console.log(action);
    //console.log(types.logout);
    switch ( action.type ) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }
    
        default:
            return state;
    }
}