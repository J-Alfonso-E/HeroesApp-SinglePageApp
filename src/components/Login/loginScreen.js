import React, {useContext} from "react";
import { useNavigate } from 'react-router-dom';
import { AuthContext } from "../../auth/authContext";
import { types } from '../../types/types';

export const LoginScreen = () => {

    const navigate = useNavigate();

    //El user se desestructura para ver que si se asigna un valor
    const {user, dispatch} = useContext(AuthContext);

    //console.log(user);

    const HandleLogin = () => {

        const action = { 
            type: types.login,
            payload: {name: "Poncho"}
        }

        const lastpath = localStorage.getItem('lastpath') || '/'

        dispatch(action);
        navigate(lastpath, { replace: true });

        //console.log(user);
        
    }

    

    return (<div>
        <h1>LoginScreen</h1>
        <hr />

        <button className="btn btn-primary" onClick={HandleLogin}>Login</button>
    </div>);
}