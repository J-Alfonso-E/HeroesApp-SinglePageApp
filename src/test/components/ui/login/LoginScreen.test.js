import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../../../auth/authContext';
import { LoginScreen } from "../../../../components/Login/loginScreen";
import { types } from '../../../../types/types';

const mockNavigate = jest.fn(); //Funcion de prueba de useNavigate, forzosamente se requiere que comience con "mock"

jest.mock('react-router-dom', () => ({ //Simulacion de la libreria de de react-router-dom
    ...jest.requireActual('react-router-dom'), //Importa la libreria de react sin cambios
    useNavigate: () => mockNavigate //Despues de que se importo se sobreescribe el hook de useNavigate
}));

describe('Pruebas en Login', () => {
    test("Snapshot del componente", () => {

        const contextValue = { user: { logged: false } }

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/login']}>
                    <LoginScreen />
                </MemoryRouter>
            </AuthContext.Provider>


        );

        expect(wrapper).toMatchSnapshot();
    });

    test('debe de realizar el dispatch y la navegacion', () => {
        const contextValue = {
            dispatch: jest.fn(),
            user: {
                logged: true,
                name: 'Poncho'
            }
        }


        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/login']}>
                    <LoginScreen />
                </MemoryRouter>
            </AuthContext.Provider>


        );

        wrapper.find('button').simulate('click');
        expect(contextValue.dispatch).toHaveBeenCalledWith({
            payload:
                { name: 'Poncho' },
            'type': types.login
        }
        ); //Verificamos que se haya llamado con el siguiente objeto
        expect(mockNavigate).toHaveBeenCalledWith('/', {replace: true}); //Verificamos que se haya llamado el hook de navegacion a una ruta especificada

        localStorage.setItem('lastpath', '/dc'); //guaramos algo en el localStorage simulando la ultima pagina visitada

        wrapper.find('button').simulate('click'); //Hacemos click en el boton de iniciar session

        expect(mockNavigate).toHaveBeenCalledWith('/dc', {replace: true}); //Verificamos que el hook de navegacion dirija al usuario a dicha pagina guardada en el localStorage

    })
});