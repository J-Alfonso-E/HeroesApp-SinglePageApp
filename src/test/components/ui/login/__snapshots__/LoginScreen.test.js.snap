// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Login Snapshot del componente 1`] = `
<div>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
